cps/src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ParallelStep.java

```diff
@@ -76,47 +107,70 @@ Callback callbackFor(String name) {
                 handler.outcomes.put(name, new Outcome(null, t));
-                checkAllDone();
+                if (handler.originalFailure == null) {
+                    handler.originalFailure = new SimpleEntry<String, Throwable>(name, t);
+                }
+                checkAllDone(true);
             }
 
-            private void checkAllDone() {
+            private void checkAllDone(boolean stepFailed) {
                 Map<String,Object> success = new HashMap<String, Object>();
-                Entry<String,Outcome> failure = null;
                 for (Entry<String,Outcome> e : handler.outcomes.entrySet()) {
                     Outcome o = e.getValue();
 
-                    if (o==null)
-                        return; // some of the results are not yet ready
+                    if (o==null) {
+                        // some of the results are not yet ready
+                        if (stepFailed && handler.failFast && ! handler.isStopSent()) {
+                            handler.stopSent();
+                            try {
+                                handler.stepExecution.stop(new FailFastException());
+                            }
+                            catch (Exception ignored) {
+                                // ignored.
+                            }
+                        }
+                        return;
+                    }
                     if (o.isFailure()) {
-                        failure= e;
+                        if (handler.originalFailure == null) {
+                            // in case the plugin is upgraded whilst a parallel step is running
+                            handler.originalFailure = new SimpleEntry<String, Throwable>(e.getKey(), e.getValue().getAbnormal());
+                        }
+                        // recorded in the onFailure
                     } else {
                         success.put(e.getKey(), o.getNormal());
                     }
                 }
-
                 // all done
-                if (failure!=null) {
+                if (handler.originalFailure!=null) {
                     // wrap the exception so that the call stack leading up to parallel is visible
-                    handler.context.onFailure(new ParallelStepException(failure.getKey(), failure.getValue().getAbnormal()));
+                    handler.context.onFailure(new ParallelStepException(handler.originalFailure.getKey(), handler.originalFailure.getValue()));
                 } else {
                     handler.context.onSuccess(success);
                 }
             }
-
+            
             private static final long serialVersionUID = 1L;
         }
 
         private static final long serialVersionUID = 1L;
     }
 
+    /** Internal exception that is only used internally to abort a parallel body in the case of a failFast body failing. */
+    private static final class FailFastException extends Exception {
```

Comment:
Users are going to see this exception, no? Don't we need to explain this like users will read it?